### PR TITLE
Add Quick View functionality on Curriculum Catalog Card

### PR DESCRIPTION
**Pull request by megcrenshaw**
This implements the Quick View button in the Curriculum Catalog Card.

The original implementation of Quick View necessitated a `<button>` tag since the action would stay on the page. But since this button now takes you to a new page, it needed an `<a>` tag –– the only way to have an `a` tag in our Button component is to add the `__useDeprecatedTag` prop. I got [confirmation](https://codedotorg.slack.com/archives/C9MB4CL07/p1682085694849349?thread_ts=1682013302.066119&cid=C9MB4CL07) that this is an okay workaround for now in order to preserve styling, and @levadadenys will take on updating our Button component to accommodate hrefs.

There is follow-up work to ensure the presence of certain fields, including the path, on the backend https://codedotorg.atlassian.net/browse/ACQ-557. For now, I am only allowing the card to show if there is a path present on the frontend. All the courses we currently show have a path, but if that were to somehow change, the entire page would crash –– so preventing that card from trying to render is a safeguard against that.

## Testing story

Unit tests reflect changes, and tested manually that I could click on the links, tab navigate and press the links with Enter, and that the screenreader reads the course offering name when on the button.

For validating the presence of link in Unit and UnitGroup, I also ran the following on production:
```
[production] dashboard > Unit.all.each.select do |unit| unit.link.nil? end
=> []
[production] dashboard > UnitGroup.all.each.select do |unit| unit.link.nil? end
=> []
```

## Follow-up work

See link above about validating the presence of certain fields.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/test/unit/templates/curriculumCatalog/CurriculumCatalogTest.jsx
Comment (TurnerRiley) [inline apps/test/unit/templates/curriculumCatalog/CurriculumCatalogTest.jsx:103]: TIL about queryByText!

**Review by TurnerRiley — APPROVED** [review]
Looks great! I did have one question about `course_offering.rb`

File: dashboard/app/models/course_offering.rb
Comment (TurnerRiley) [inline dashboard/app/models/course_offering.rb:235]: What does adding `summarize_for_catalog` (and replacing `serialize` with `summarize_for_catalog` in `curriculum_catalog_controller.rb`) do here? I saw above in the changes to `curriculumCatalogShape.jsx`, several props were removed (similar to `summarize_for_catalog` having fewer fields than `serialize`) since they aren't being used which makes sense, but won't we need to use them in the future?

Comment (TurnerRiley) [reply dashboard/app/models/course_offering.rb:235]: Oh sorry, I reposted this comment highlighting the full text so it was clearer what I was pointing to. I remember working on a PR to add `marketing_initiative` as a prop recently because it was going ot be needed so I imagine the others will be needed as well. I agree, we should probably keep it minimal but if it's already there I feel like we might as well keep it until it's useful (assuming it will be useful soon) rather than remove it only to add it back in shortly.

Comment (megcrenshaw) [reply dashboard/app/models/course_offering.rb:235]: Good call –– I'm under the impression that we don't need them, but I'm realizing device compatibility is needed. Are there others that will be needed?

Part of me wonders if we should add attributes as we go (keeping it as minimal as possible until we need more), but if it's clear we really will need every single field, I agree it doesn't make sense to create a separate category.

Comment (megcrenshaw) [reply dashboard/app/models/course_offering.rb:235]: The big thing is now we have an attribute (course_version_path) that is not a serialized field. Do you think that attribute should be serialized too?

Comment (megcrenshaw) [reply dashboard/app/models/course_offering.rb:235]: Oh I'm seeing device compatibility is still in there 😅

Comment (TurnerRiley) [reply dashboard/app/models/course_offering.rb:235]: I think it would be good to also have it be serialized if possible! Unless if there's a reason you're feeling it shouldn't be

Comment (megcrenshaw) [reply dashboard/app/models/course_offering.rb:235]: It feels risky to have it serialized because if the latest published version of the course offering changes, the link would change here too –– and then we'd have two sources of truth (one in the DB and one in the JSON). What is your reasoning for ensuring it's serialized?

Comment (TurnerRiley) [reply dashboard/app/models/course_offering.rb:235]: Ohhhhh I see, yes you're right good call. Okay I feel comfortable with this then! Thank you for walking me through that!

**Comment by davidsbailey** [comment]
> There is follow-up work to ensure the presence of certain fields, including the path, on the backend https://codedotorg.atlassian.net/browse/ACQ-557. For now, I am only allowing the card to show if there is a path present on the frontend. All the courses we currently show have a path, but if that were to somehow change, the entire page would crash –– so preventing that card from trying to render is a safeguard against that.

pre-review comment: this makes sense to me while path is not required on the backend. once path is required on the backend, it seems like it could be ok (even desirable) for the frontend page to crash (or at least hit a proptype warning), since this would indicate a programming error of some kind, and we would want our tests to fail.

File: apps/src/templates/curriculumCatalog/CurriculumCatalog.jsx
Comment (davidsbailey) [inline apps/src/templates/curriculumCatalog/CurriculumCatalog.jsx:86]: per other comment, I think you should consider removing the `course_version_path` check here (and mark prop as required) as part of  https://codedotorg.atlassian.net/browse/ACQ-557.

File: dashboard/app/models/course_offering.rb
Comment (davidsbailey) [inline dashboard/app/models/course_offering.rb:85]: 💯

File: dashboard/app/models/course_offering.rb
Comment (davidsbailey) [reply dashboard/app/models/course_offering.rb:235]: +1 Meg, thanks for explaining! This makes sense to me

File: dashboard/test/models/course_offering_test.rb
Comment (davidsbailey) [inline dashboard/test/models/course_offering_test.rb:258]: nice test! I'm glad to see the factory is working, though it is still sad that you have to know that course version is the factory to use if you want it to create the other stuff for you.

I think it would be helpful to add an similar test where all of the content roots are `with_unit`, and update the name of this test to make it clear that it is hitting the unit group case, since the default factory behavior is not obvious here.

**Review by davidsbailey — APPROVED** [review]
Great work Meg! LGTM after adding 1 more test case

File: dashboard/test/models/course_offering_test.rb
Comment (megcrenshaw) [reply dashboard/test/models/course_offering_test.rb:258]: Yeah I hear that –– sounds good, will do in a follow-up PR!

File: apps/src/templates/curriculumCatalog/CurriculumCatalog.jsx
Comment (megcrenshaw) [reply apps/src/templates/curriculumCatalog/CurriculumCatalog.jsx:86]: Yes agreed. UI tests will cover if the page is crashing.

File: dashboard/test/models/course_offering_test.rb
Comment (megcrenshaw) [reply dashboard/test/models/course_offering_test.rb:258]: Oh wait for some reason I thought I had merged this –– I'll do this in this PR!

Comment (megcrenshaw) [reply dashboard/test/models/course_offering_test.rb:258]: Added a test to address `unit`. For getting the course version factory to do what we're wanting, I had tried `association :course_version` before I realized that the association goes the other way. I think I'd need to do something like `course_version {create :course_version}` –– I can think about this in the follow-up validation PR.

Comment (davidsbailey) [reply dashboard/test/models/course_offering_test.rb:258]: I think I am missing a piece here. which factory did you mean adding `association :course_version` and `course_version {create :course_version}` to? and/or did you mean `course_offering` in any of the places it says `course_version`?

Comment (megcrenshaw) [reply dashboard/test/models/course_offering_test.rb:258]: I think I meant `course_version` for all of these. A course offering `has many` course versions. Inside the course version factory, we do `association :course_offering`.

I had originally tried to do `association :course_version` inside the course offering factory, but the association goes the other way. But for a course offering to correctly run `course_offering.course_versions` I'm wondering if we need to add course_versions explicitly, like `course_version {create :course_version}`, similar to this:
https://github.com/code-dot-org/code-dot-org/blob/6519c6c92a0fc29f783d7ef5cfd3dd65c9a998f6/dashboard/test/factories/pd_factories.rb#L517

That's for one, but for the has_many association, it looks similar but with a loop, e.g.
https://thoughtbot.com/blog/aint-no-calla-back-bot

Comment (davidsbailey) [reply dashboard/test/models/course_offering_test.rb:258]: your goal is to do something like `create :course_offering, :with_units` right? 

> I had originally tried to do association :course_version inside the course offering factory, but the association goes the other way.

I'm not sure I understand all of the paths you have explored, but this works for me:
```diff
diff --git a/dashboard/test/factories/factories.rb b/dashboard/test/factories/factories.rb
index dd3e14249a7..40c9ee0ecf2 100644
--- a/dashboard/test/factories/factories.rb
+++ b/dashboard/test/factories/factories.rb
@@ -5,6 +5,13 @@ FactoryBot.define do
     sequence(:key, 'a') {|c| "bogus-course-offering-#{c}"}
     sequence(:display_name, 'a') {|c| "bogus-course-offering-#{c}"}
     assignable {true}
+
+    trait :with_units do
+      after(:create) do |course_offering|
+        create(:course_version, :with_unit, course_offering: course_offering)
+        create(:course_version, :with_unit, course_offering: course_offering)
+      end
+    end
   end
```

Comment (megcrenshaw) [reply dashboard/test/models/course_offering_test.rb:258]: Ah, this is great! Thanks –– updated the factories to do this

**Pull request merged**
